Add option to set the ellipsoid to use for calculations with qgis_process tool

nyalldawson opened the pull request:
- Add explicit settings for ellipsoid/distance unit/area unit to QgsProcessingContext, and prefer to use these instead of trying to get the settings from a QgsProject (which isn't available from the standalone tool).
- Expose a new argument to allow users to specify "--ELLIPSOID=xxxx" when running algorithms via the qgis_process tool.
- Flags algorithms which make no sense to run outside of the GUI environment (such as "select by ..." algorithms), and hides/prevents them from running via qgis_process.
- Also warns when running deprecated or algorithms with known issues from qgis_process
- Flags algorithms which require a valid project to execute so that errors are raised when running these from qgis_process, unless the new "--PROJECT_PATH=xxxx" argument is specified and pointing to a valid project path.

Fixes #37988

nyalldawson committed:
Add explicit settings for ellipsoid, distance and area units to QgsProcessingContext
to avoid having to always retrieve these via the context's project (which may
not be available in some circumstances)

Refs #37988

nyalldawson committed:
Use direct getters for ellipsoid/units in processing algs

Refs #37988

nyalldawson committed:
[processing] Add option to set the ellipsoid to use for calculations
when running qgis_process

Appending "--ELLIPSOID=xxxx" to the argument list allows users to explicitly
set the ellipsoid to use for calculations.

Fixes #37988

agiudiceandrea commented on python/plugins/processing/algs/qgis/ExportGeometryInfo.py:
```diff
@@ -138,7 +138,7 @@ def processAlgorithm(self, parameters, context, feedback):
-            self.distance_area.setEllipsoid(context.project().ellipsoid())
+            self.distance_area.setEllipsoid(context.ellipsoid())
         elif method == 1:
             coordTransform = QgsCoordinateTransform(source.sourceCrs(), context.project().crs(), context.project())
```
Hi @nyalldawson, this algorithm also needs `context.project().crs()` and `context.project()`

`context.project().crs()` is also needed for example by [native:networkanalysis](https://github.com/qgis/QGIS/blob/master/src/analysis/processing/qgsalgorithmnetworkanalysisbase.cpp#L116)

What about some other algs that use the whole context.project() like for example native:rasterize native:loadlayer, native:setprojectvariable, native:stylefromproject, native:bookmarkstolayer, native:categorizeusingstyle, ... ?

nyalldawson replied:
Yes, it's only a partial fix but addresses the immediate ellipsoid handling need. 

I guess we also need a --project=... argument , and then make sure this is specified if an algorithm requires it. That said, "export geometry info" is a generally badly behaved algorithm and needs a rewrite (in C++). And some of those other algorithms have no real meaning outside of qgis and should be hidden entirely (such as the "select by...." algorithms). I'll add a flag to hide algorithms from the standalone tool to handle this.

agiudiceandrea replied:
Thank you very much!

nyalldawson replied:
Ok, I've updated this PR with the extra logic now...

nyalldawson committed:
Fix test

nyalldawson committed:
[processing] Add a flag for algorithms to indicate that they should
not be available from the standalone qgis_process tool (e.g. "select by "
algorithms, which have no meaning outside of a GUI application)

These algorithms are hidden from the algorithm list for qgis_process
and cannot be run by the tool

nyalldawson committed:
[processing] Flag a bunch of algorithms to hide them from qgis_process

nyalldawson committed:
Warn on running deprecated algorithms or algorithms with known issues from qgis_process

nyalldawson committed:
Add a processing algorithm flag to indicate that an algorithm requires
a valid project to be set in the context in order to execute

And flag a bunch of algorithms which require this

nyalldawson committed:
Gracefully handle no project available in context in algorithms
where a project is only required in some circumstances

nyalldawson committed:
Use correct coordinate transform context in Raster Calculator algorithm

nyalldawson committed:
[processing] Allow specifying a project path to load when running
algorithms via qgis_process

Allows execution of algorithms which require a project, e.g. Rasterize

Also show graceful errors/help text when attempting to run an algorithm
which requires a project but one is not specified

nyalldawson commented:
Unrelated test failure

Gustry commented:
Nice addition !
Can you add the new project option in the PR description (for changelog and doc)? I can't edit. Thanks

nyalldawson commented:
@Gustry done

agiudiceandrea commented:
@nyalldawson did you forget to mention the "**--AREA_UNIT**=" and "**--DISTANCE_UNIT**=" parameters in the PR description and in the usage description? Or are they not intended to be used?
